Fixed #32425 -- Fixed adding nullable field with default on MySQL.

baidoosik opened the pull request:
ticket link: https://code.djangoproject.com/ticket/32425#ticket

Hi, team. I wanna fix some picky issue. 
I suggest below code for solving this problem. (same code creates different schemas)

but It has duplicate code with base/schema.py _alter_column_default_sql.

We can do like this. but it need to one more SQL. 
For mysql users, performance is not good, but better than schema changes.

mysql/schema.py

```
def _alter_column_default_sql(self, model, old_field, new_field, drop=False):
      super()._alter_column_default_sql(model, old_field, new_field, drop)

      if drop and new_field.null:
          new_db_params = new_field.db_parameters(connection=self.connection)
          new_default = self.effective_default(new_field)
          default = self._column_default_sql(new_field)
          params = [new_default]
          sql = self.sql_alter_column_default_null
          return (
              sql % {
                  'column': self.quote_name(new_field.column),
                  'type': new_db_params['type'],
                  'default': default,
              },
              params,
          )

```

baidoosik commented:
@charettes Thank you for reviewing the suggestion.
I updated the code according to your guide.
Could you check this? ​https://github.com/django/django/pull/13982
If this is proper to what you're saying, I will add testcases.
And I wanna check where is it right to put the testcases.
I guess tests/backends/test_schema.py.

charettes commented on django/db/backends/base/schema.py:
```diff
@@ -62,6 +62,8 @@ class BaseDatabaseSchemaEditor:
     sql_alter_column_not_null = "ALTER COLUMN %(column)s SET NOT NULL"
     sql_alter_column_default = "ALTER COLUMN %(column)s SET DEFAULT %(default)s"
     sql_alter_column_no_default = "ALTER COLUMN %(column)s DROP DEFAULT"
+    sql_alter_column_no_default_null = "ALTER COLUMN %(column)s SET DEFAULT NULL"
+
```
```suggestion
    sql_alter_column_no_default_null = sql_alter_column_no_default
```

And then override it in `mysql.schema.SchemaEditor.sql_alter_column_no_default_null = "ALTER COLUMN %(column)s SET DEFAULT NULL"`

baidoosik replied:
@charettes Okay i understood! Big thanks!

charettes reviewed (commented):
@baidoosik the only thing missing here is that you'll want to only use `SET DEFAULT NULL` on MySQL. See comment below.

> And I wanna check where is it right to put the testcases. I guess tests/backends/test_schema.py.

yeah that seems about right, here's a minimal test that reproduced your issue

```diff
diff --git a/tests/schema/tests.py b/tests/schema/tests.py
index b994c252d969..e8a2e743d9da 100644
--- a/tests/schema/tests.py
+++ b/tests/schema/tests.py
@@ -3095,6 +3095,30 @@ def test_add_field_default_dropped(self):
             if connection.features.can_introspect_default:
                 self.assertIsNone(field.default)
 
+    def test_add_field_default_nullable(self):
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+        # Add new nullable CharField with a default
+        new_field = CharField(max_length=15, blank=True, null=True, default='surname')
+        new_field.set_attributes_from_name("surname")
+        with connection.schema_editor() as editor:
+            editor.add_field(Author, new_field)
+        Author.objects.create(name='Anonymous1')
+        # Ensure field is still nullable
+        with connection.cursor() as cursor:
+            cursor.execute("SELECT surname FROM schema_author;")
+            item = cursor.fetchall()[0]
+            self.assertIsNone(item[0])
+            field = next(
+                f for f in connection.introspection.get_table_description(cursor, "schema_author")
+                if f.name == "surname"
+            )
+            self.assertIs(True, field.null_ok)
+            # And that the default is no longer set in the database.
+            if connection.features.can_introspect_default:
+                self.assertIsNone(field.default)
+
     def test_alter_field_default_dropped(self):
         # Create the table
         with connection.schema_editor() as editor:
```

baidoosik commented:
@charettes I updated your review. and I think the testcase you suggest is enough for this case(small changes). And i modified  a few things from your suggestion. The definition was slightly different according to db, so I modified that part.
If you think to need more test cases or anything, tell me!

charettes commented on tests/schema/tests.py:
```diff
@@ -3095,6 +3095,30 @@ def test_add_field_default_dropped(self):
+                f for f in connection.introspection.get_table_description(cursor, "schema_author")
+                if f.name == "surname"
+            )
+            self.assertTrue(field.null_ok in (1, True))
```
Ah I think you could use `assertTrue` without issues as both `1` and `True` are truthy.

```suggestion
            self.assertTrue(field.null_ok)
```

baidoosik commented:
@charettes  I updated your review! Thanks!

felixxm commented:
@baidoosik Thanks :+1: Welcome aboard :boat: 

I fixed crash on Oracle.

felixxm commented:
buildbot, test on oracle.

baidoosik commented:
@felixxm Big Thanks! I will try to join the community through this chance!

felixxm commented:
buildbot, test on oracle.
